### PR TITLE
Intra EVC inactive state and leftover flows solved.

### PR DESCRIPTION
**Pull request by Alopalao**
Closes #719 

### Summary

See updated changelog file and/or add any other summarized helpful information for reviewers

### Local Tests
Performed 2 tests.
1.
Started kytos without `-E` option
Created an intra EVC
Enabled switch from recently create EVC UNIs
Check if the flows from the intra EVC are installed

2.
Disabled the intra EVC so flows are uninstalled
Disabled switch from UNIs
Enabled intra EVC, check that flows are not installed
Enable switch from UNIs

3.
Also performed the test from this [comment](https://github.com/kytos-ng/mef_eline/pull/721#pullrequestreview-4462864029).

### End-to-End Tests
[In progress](https://github.com/kytos-ng/kytos-end-to-end-tests/pull/457)
```
+ python3 -m pytest tests/ --reruns 2 -r fEr
============================= test session starts ==============================
platform linux -- Python 3.11.2, pytest-8.4.2, pluggy-1.6.0
rootdir: /
configfile: pytest.ini
plugins: timeout-2.2.0, rerunfailures-13.0, asyncio-1.1.0, anyio-4.3.0
asyncio: mode=Mode.AUTO, asyncio_default_fixture_loop_scope=None, asyncio_default_test_loop_scope=function
collected 309 items

tests/test_e2e_01_kytos_startup.py ..                                    [  0%]
tests/test_e2e_05_topology.py ...................                        [  6%]
tests/test_e2e_06_topology.py .....                                      [  8%]
tests/test_e2e_08_topology.py .                                          [  8%]
tests/test_e2e_10_mef_eline.py ..........ss.....x....................... [ 22%]
.                                                                        [ 22%]
tests/test_e2e_11_mef_eline.py ........                                  [ 24%]
tests/test_e2e_12_mef_eline.py .....Xx..                                 [ 27%]
tests/test_e2e_13_mef_eline.py .......................ssss.............. [ 41%]
.                                                                        [ 41%]
tests/test_e2e_14_mef_eline.py ......                                    [ 43%]
tests/test_e2e_15_mef_eline.py ......                                    [ 45%]
tests/test_e2e_16_mef_eline.py ..                                        [ 45%]
tests/test_e2e_17_mef_eline.py .....                                     [ 47%]
tests/test_e2e_18_mef_eline.py .......                                   [ 49%]
tests/test_e2e_20_flow_manager.py .............................          [ 59%]
tests/test_e2e_21_flow_manager.py ...                                    [ 60%]
tests/test_e2e_22_flow_manager.py ...............                        [ 65%]
tests/test_e2e_23_flow_manager.py ..............                         [ 69%]
tests/test_e2e_30_of_lldp.py ......                                      [ 71%]
tests/test_e2e_31_of_lldp.py ....                                        [ 72%]
tests/test_e2e_32_of_lldp.py ...                                         [ 73%]
tests/test_e2e_40_sdntrace.py ................                           [ 78%]
tests/test_e2e_41_kytos_auth.py ........                                 [ 81%]
tests/test_e2e_42_sdntrace.py ..                                         [ 82%]
tests/test_e2e_50_maintenance.py ...............................         [ 92%]
tests/test_e2e_60_of_multi_table.py .....                                [ 93%]
tests/test_e2e_70_kytos_stats.py .........                               [ 96%]
tests/test_e2e_80_pathfinder.py ss......                                 [ 99%]
tests/test_e2e_90_kafka_events.py .                                      [ 99%]
tests/test_e2e_95_telemtry_int.py s                                      [100%]
```

### CODE REVIEW

File: models/evc.py [hunk 1]
```diff
@@ -1748,6 +1753,14 @@ def get_endpoint_by_id(
             return link.endpoint_a
         return link.endpoint_b
 
+    def intra_evc_needs_redeployment(self) -> bool:
+        """Determine whether an intra-switch EVC needs redeployment."""
+        if not self.is_intra_switch():
+            return False
```
Comment (viniarck) [inline models/evc.py:1776]: You can make the code intention easier to follow by:

Moving `if not self.last_deployed_at: return True` as an early return.

That makes immediately clear that if it's not set it needs redeployment, currently whoever is reading needs to first understand what `flows_removed` bool result means and for the prior return it's another conditional despite the logical or result being correct.

File: models/evc.py [hunk 2]
```diff
@@ -1748,6 +1753,14 @@ def get_endpoint_by_id(
...
+        flows_removed = bool(self.flow_removed_at) and \
```
Comment (viniarck) [inline models/evc.py]: 1) `self.flow_removed_at` state was meant to capture the last time where a flow was removed from a flow removed message, it's not too far away from what you want here, but keep in mind it depends on the switch response and it's threaded, so that can result in late execution in worst case scenarios, so you end up with a specific extreme corner case where `flow_removed_at` took so long to arrive before the last_deployed_at then the order you expect here wouldn't enter here, it'd be very rare but it's possible especially if that thread handler is very slow and taking a while to execute. That said, you might want to introduce something else similar to `last_deployed_at` but for removal (`last_removed_at` or `path_removed_at` or something better named) and then have the controller set it then you can rely on it without other major Switch FlowMods and threaded related delays.

2) It'll also enhance readability by also making sure `self.last_deployed_at` is truthy before comparing with `self.flow_removed_at`, same idea with your `bool` cast, although just checking for a truthy value here would suffice too without `bool`

Think about see what else you can improved here

Comment (viniarck) [reply models/evc.py]: Finally, it's good to also include `last_deployed_at` in the openapi.yml spec Circuit response, there are other fields that we left behind but it should be there too for docs, and for also clients to know they can use and read that value

**Review by viniarck — CHANGES_REQUESTED** [review]
Overall it's in the right direction @Alopalao, there are minor threaded state potential issues that need to be sorted out though.

File: models/evc.py (hunk 1 above)
Comment (Alopalao) [reply models/evc.py:1776]: Ok, now returning early if `not self.last_deployed_at` is true

File: models/evc.py (hunk 2 above)
Comment (Alopalao) [reply models/evc.py]: Added the missing values to `openapi.yml` and encapsulating all the return with bool, it was returning `None`.
Also implemented `last_remove_at` to keep track when `current_path` was removed.

File: openapi.yml [hunk 3]
```diff
@@ -718,6 +718,34 @@ components:
...
+          description: "Time when the flow was removed from the switch by flow_manager."
+        last_deployed_at:
+          type: string
+          format: date-time
+          description: "Time when the circuit was last deployed and flows were sent to be installed."
+        last_removed_at:
+          type: string
+          format: date-time
+          description: "Time when the circuit current path was last removed and flows were sent to be removed."
+        updated_at:
```
Comment (viniarck) [inline openapi.yml:743]: Thanks for all the OpenAPI doc improvements @Alopalao

**Review by viniarck — CHANGES_REQUESTED** [review]
@Alopalao pretty neat how it turned out, very easy to follow too. 

There's still one corner case to handle when a EVC changes from intra to inter though: 

- It can leave installed flows behind when updating both UNIs to different switches, say you have S1:1 and S1:2, and then you update as an inter to S2:1 and S3:1, S1 UNI flows will be left behind. Example using ring topo:

```
❯ echo '{ "name": "intra_epl", "dynamic_backup_path": true, "queue_id": 1, "uni_a": { "interface_id": "00:00:00:00:00:00:00:01:1" }, "uni_z": { "interface_id": "00:00:00:00:00:00:00:01:2" } }' | http http://
127.0.0.1:8181/api/kytos/mef_eline/v2/evc/
HTTP/1.1 201 Created
content-length: 47
content-type: application/json
date: Tue, 09 Jun 2026 22:49:11 GMT
server: uvicorn

{
    "circuit_id": "7fa4e863c9e74e",
    "deployed": true
}
``` 

```
rs0 [direct: primary] napps> db.flows.find({"state": "installed", "flow.owner": "mef_eline"})
[
  {
    _id: 'a3554d844abe3c9ee3bc141ca7af88dd',
    flow: {
      table_id: 0,
      owner: 'mef_eline',
      table_group: 'epl',
      priority: 10000,
      cookie: Decimal128('12285719626503546702'),
      idle_timeout: 0,
      hard_timeout: 0,
      match: { in_port: 1 },
      actions: [
        { action_type: 'set_queue', queue_id: 1 },
        { action_type: 'output', port: 2 }
      ]
    },
    flow_id: '216ec750f2bdc5c0c69124c5bc4c1dbc',
    id: 'a3554d844abe3c9ee3bc141ca7af88dd',
    inserted_at: ISODate('2026-06-09T22:49:12.283Z'),
    state: 'installed',
    switch: '00:00:00:00:00:00:00:01',
    updated_at: ISODate('2026-06-09T22:49:12.298Z')
  },
  {
    _id: 'a50edc47b42f556ce118878b4c1ec22e',
    flow: {
      table_id: 0,
      owner: 'mef_eline',
      table_group: 'epl',
      priority: 10000,
      cookie: Decimal128('12285719626503546702'),
      idle_timeout: 0,
      hard_timeout: 0,
      match: { in_port: 2 },
      actions: [
        { action_type: 'set_queue', queue_id: 1 },
        { action_type: 'output', port: 1 }
      ]
    },
    flow_id: '0b8df2be6cae3b083778d42fe6334f36',
    id: 'a50edc47b42f556ce118878b4c1ec22e',
    inserted_at: ISODate('2026-06-09T22:49:12.283Z'),
    state: 'installed',
    switch: '00:00:00:00:00:00:00:01',
    updated_at: ISODate('2026-06-09T22:49:12.298Z')
  }
]
rs0 [direct: primary] napps> 

```

Now, let's update both UNIs but none of the UNIs will use S1:

```
❯ echo '{ "uni_a": { "interface_id": "00:00:00:00:00:00:00:02:1", "tag": {"tag_type": "vlan", "value": 100} }, "uni_z": { "interface_id": "00:00:00:00:00:00:00:03:1", "tag": {"tag_type": "vlan", "value": 100
} } }' | http PATCH 'http://localhost:8181/api/kytos/mef_eline/v2/evc/7fa4e863c9e74e'
```

Notice this UNI flow on S1 is still installed;

```
rs0 [direct: primary] napps> db.flows.find({"state": "installed", "flow.owner": "mef_eline", "switch": "00:00:00:00:00:00:00:01", "flow.match.in_port": 1})
[
  {
    _id: 'a3554d844abe3c9ee3bc141ca7af88dd',
    flow: {
      table_id: 0,
      owner: 'mef_eline',
      table_group: 'epl',
      priority: 10000,
      cookie: Decimal128('12285719626503546702'),
      idle_timeout: 0,
      hard_timeout: 0,
      match: { in_port: 1 },
      actions: [
        { action_type: 'set_queue', queue_id: 1 },
        { action_type: 'output', port: 2 }
      ]
    },
    flow_id: '216ec750f2bdc5c0c69124c5bc4c1dbc',
    id: 'a3554d844abe3c9ee3bc141ca7af88dd',
    inserted_at: ISODate('2026-06-09T22:49:12.283Z'),
    state: 'installed',
    switch: '00:00:00:00:00:00:00:01',
    updated_at: ISODate('2026-06-09T22:49:12.298Z')
  }
]
```

Let's cover this with both a unit and e2e test

**Comment by Alopalao** [comment]
> There's still one corner case to handle when a EVC changes from intra to inter though:

Patched this behavior, now EVC has `leftover_switch` which is saved when the EVC is updated and removed when the current path is removed. Also added new e2e test in the same [PR](https://github.com/kytos-ng/kytos-end-to-end-tests/pull/457) for this one.
